plugins/producer_plugin/test/test_snapshot_scheduler.cpp

```
@@ -0,0 +1,140 @@
+                  {"test", "--data-dir", temp.c_str(), "--config-dir", temp.c_str(),
+                   "-p", "eosio", "-e", "--max-transaction-time", "475", "--disable-subjective-billing=true"};
+            appbase::app().initialize<chain_plugin, producer_plugin>(argv.size(), (char**) &argv[0]);
+            appbase::app().startup();
+            plugin_promise.set_value(
+                  {appbase::app().find_plugin<producer_plugin>(), appbase::app().find_plugin<chain_plugin>()});
+            appbase::app().exec();
+         });
+
+         auto [prod_plug, chain_plug] = plugin_fut.get();
+         std::deque<block_state_ptr> all_blocks;
+         std::promise<void> empty_blocks_promise;
+         std::future<void> empty_blocks_fut = empty_blocks_promise.get_future();
+         auto ab = chain_plug->chain().accepted_block.connect([&](const block_state_ptr& bsp) {
+            static int num_empty = std::numeric_limits<int>::max();
+            all_blocks.push_back(bsp);
+            if(bsp->block->transactions.empty()) {
+               --num_empty;
+               if(num_empty == 0) empty_blocks_promise.set_value();
+            } else {// we want a few empty blocks after we have some non-empty blocks
+               num_empty = 10;
+            }
+         });
+         auto pp = appbase::app().find_plugin<producer_plugin>();
+         auto bs = chain_plug->chain().block_start.connect([&pp](uint32_t bn) {
+            // catching pending snapshot
+            if (pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots && pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots->size()==1) {
+               // lets check the head block num of it, it should be 8 + 1 = 9
+               // this means we are getting a snapshot for correct block # as well
+               BOOST_CHECK_EQUAL(9, pp->get_snapshot_requests().snapshot_requests.begin()->pending_snapshots->begin()->head_block_num);   
```

We never get here, the `pending_snapshots` vector is always empty.

---

You are right, we get there. I'm not understanding why the check doesn't fail when we take a snapshot at block 3:

`snapshot scheduler creating a snapshot from the request [start_block_num:3, end_block_num=3, block_spacing=2], height=4`

---

It fails for me in this case and fails hard, but not in this check if I modify sri1. Problem with start=end request that it will be removed immediately after execution so this code checking for "9" wont be reached